### PR TITLE
Ticket3409 separator stability

### PR DESCRIPTION
**Pull request by Alistair-McGann-Tessella**
### Description of work
Described in detail here: https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Muon-Separator-Power-Supply#separator-stability

### To test

https://github.com/ISISComputingGroup/IBEX/issues/3409

### Acceptance criteria

- [x] Unstable data can be correctly identified and logged when hardware is sampled at a reasonable rate
- [x] A PV exists which shows the time spend outside of stable regions within an arbitrary length rolling buffer

### CODE REVIEW

File: separatorSup/separator_stability.db [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    field(EGU, "kV")
+    field(VAL, "$(VLOWERLIM)")
+}
+
+record(ai, "$(P)CURR:STEADY")
+{
+    field(DESC, "Nominal current value")
+    field(SCAN, "Passive")
+    field(EGU, "mA")
+    field(VAL, "$(ISTEADY)")
```
Comment (LiamPanchaud) [inline separatorSup/separator_stability.db:38]: I think a steady current is 0A (±a little bit). I think this is becasue when I is 0 the separator is sitting at a steady x kV, where x is whatever V they want it at, The only time I goes up is when it's charging, which doesn't cound as stable. I think ISTEADY will always be 0mA. I may be wrong though, tell me if you know something I don't

Comment (kjwoodsISIS) [reply separatorSup/separator_stability.db:38]: Yes, ISTEADY will be zero (or close to it) - see [Muon FE Separator Criteria](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/Muon%20FE/Muon%20FE%20Separator%20Criteria.docx).  ISTEADY can be set, by configuration, to be zero.

File: separatorSup/separator.c [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+#include <epicsExport.h>
+
+#include "separator.h"
+
+static long parse_stability(aSubRecord *prec)
+{
+    return parse_stability_impl(prec);
+}
+
+epicsRegisterFunction(parse_stability);
```
Comment (ghost) [inline separatorSup/separator.c]: Needs a new line.

File: separatorSup/separator.dbd [hunk 3]
```diff
@@ -0,0 +1 @@
+function(parse_stability)
```
Comment (ghost) [inline separatorSup/separator.dbd:1]: Maybe worth renaming the dbd file to something more descriptive, e.g. `separator_asub_function.dbd`.

File: separatorSup/separator.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern long parse_stability_impl(aSubRecord *prec);
+// extern long another_function_name_impl(aSubRecord *prec);
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (ghost) [inline separatorSup/separator.h]: New line at EOF.

File: separatorSup/separator.h [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#ifdef __cplusplus
```
Comment (ghost) [inline separatorSup/separator.h:1]: Worth wrapping the header file in a [C++ header guard](https://stackoverflow.com/questions/4767068/header-guards-in-c#4767103).

File: separatorSup/separator_impl.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
+#include <string.h>
```
Comment (ghost) [inline separatorSup/separator_impl.cpp:1]: Are all of these dependencies needed?

**Review by ghost — CHANGES_REQUESTED** [review]
- When I tested the IOC with the DAQ readings with stability limits set below the voltage level it did not show that the voltage is unstable. I feel it should do this. When the stability limits are set far above the voltage readings, it should also show that the voltage is unstable.
- The support module is getting sufficiently large that it may be worth moving the `SRC`, `db`, `dbd` files into separate folders.
- Also, it may be worth using cxxtest to test parts of the aSub record (our version cxxtest can be found at https://github.com/ISISComputingGroup/EPICS-cxxtest and is added to libraries by https://github.com/ISISComputingGroup/EPICS-libraries/pull/3).

File: separatorSup/separator_impl.cpp (hunk 6 above)
Comment (Alistair-McGann-Tessella) [reply separatorSup/separator_impl.cpp:1]: Yes, most of these are either the libraries which link EPICS to C++ or standard C++ libraries.

File: separatorSup/separator.h (hunk 5 above)
Comment (Alistair-McGann-Tessella) [reply separatorSup/separator.h:1]: I'm not sure if that would be useful in this case. There aren't any include statements here, so is the risk of having multiple definitions of something not quite low?

File: separatorSup/separator.dbd (hunk 3 above)
Comment (Alistair-McGann-Tessella) [reply separatorSup/separator.dbd:1]: This is remarkably fiddly to change, I agree it should be more descriptive but I don't think this is something that should be fixed right now. I did have a look at this, but I'm concerned that will take longer than I'm comfortable spending on it

**Review by ghost — COMMENTED** [review]
When testing the separator with the DAQ in G39, the unstable time hits a limit at about 70-80% of buffer size and stays there despite the stability limits being set such that all values are unstable.

**Comment by Alistair-McGann-Tessella** [comment]
I think this issue is an artefact of how the buffer is being filled. This is done by adding up the number of samples that are outside the stable region received by the device every second. If this record receives data while the other buffer record is getting processed then this data will be lost. Data loss can also happen because of hardware and network problems.

The 'pre-buffer' PV `_ADDCOUNTS` was implemented to alleviate this, as well as make sure that our time doesn't get out of sync. This is the most robust architecture that we could come up with to solve this problem.

I can think of a couple of ways to mitigate this problem, though:
1. Make the number of data parcels handled by the records every second 1 by making FREQ=NELM.
2. Pass data parcels that are bigger than 1 second by making NELM > FREQ. This seems to work quite well, sometimes has the odd effect of having the unstable time greater than the buffer size, but this seems to prevent data loss.
3. Make the size of the dropped parcels smaller, making their effect on the summed buffer smaller (Make NELM << FREQ). I tried this and it looks like it makes things worse, possibly because of the overhead time of the record processing, and maybe more dropped data.

**Review by ghost — APPROVED** [review]
I accept Alistair's reasoning for why there are problems in the extreme case when the current is always unstable and that in practice this will be good enough. There is a note on the wiki page outlining this issue.

**Pull request merged**